fq.go

```diff
@@ -0,0 +1,168 @@
+}
+
+func (q *FQScheduler) getvirtualtimeratio() float64 {
+	NEQ := 0
+	reqs := 0
+	for _, queue := range q.queues {
+		reqs += queue.RequestsExecuting
+		// It might be best to delete this line. If everything is working
+		//  correctly, there will be no waiting packets if reqs < C on current
+		//  line 85; if something is going wrong, it is more accurate to say
+		// that virtual time advanced due to the requests actually executing.
+
+		// reqs += len(queue.Packets)
+		if len(queue.Packets) > 0 || queue.RequestsExecuting > 0 {
+			NEQ++
+		}
+	}
+	// no active flows, virtual time does not advance (also avoid div by 0)
+	if NEQ == 0 {
+		return 0
+	}
+	return min(float64(reqs), float64(C)) / float64(NEQ)
+}
+
+func (q *FQScheduler) updateTime(packet *Packet, queue *Queue) {
+	// When a request arrives to an empty queue with no requests executing
+	// len(queue.Packets) == 1 as enqueue has just happened prior (vs  == 0)
+	if len(queue.Packets) == 1 && queue.RequestsExecuting == 0 {
+		// the queue’s virtual start time is set to getVirtualTime().
+		queue.virstart = q.getVirtualTime()
+	}
+}
+
+// FinishPacketAndDequeue is a convenience method used using the FQScheduler
+// at the concurrency limit
+func (q *FQScheduler) FinishPacketAndDeque(p *Packet) (*Packet, bool) {
+	q.FinishPacket(p)
+	return q.Dequeue()
+}
+
+// FinishPacket is a callback that should be used when a previously dequeud packet
+// has completed it's service.  This callback updates imporatnt state in the
+//  FQScheduler
+func (q *FQScheduler) FinishPacket(p *Packet) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	q.synctime()
+	S := q.clock.Since(p.startTime).Seconds()
+
+	// When a request finishes being served, and the actual service time was S,
+	// the queue’s virtual start time is decremented by G - S.
+	q.queues[p.queueidx].virstart -= G - S
+
+	// request has finished, remove from requests executing
+	q.queues[p.queueidx].RequestsExecuting--
+}
+
+// Dequeue dequeues a packet from the fair queuing scheduler
+func (q *FQScheduler) Dequeue() (*Packet, bool) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	q.synctime()
+	queue := q.selectQueue()
+
+	if queue == nil {
+		return nil, false
+	}
+	packet, ok := queue.Dequeue()
+
+	if ok {
+		// When a request is dequeued for service -> q.virstart += G
+		queue.virstart += G
+
+		packet.startTime = q.clock.Now()
+		// request dequeued, service has started
+		queue.RequestsExecuting++
+	}
+	return packet, ok
+}
+
+func (q *FQScheduler) roundrobinqueue() int {
+	q.robinidx = (q.robinidx + 1) % len(q.queues)
+	return q.robinidx
+}
+
+func (q *FQScheduler) selectQueue() *Queue {
+	minvirfinish := math.Inf(1)
+	var minqueue *Queue
+	var minidx int
+	for range q.queues {
+		idx := q.roundrobinqueue()
+		queue := q.queues[idx]
+		if len(queue.Packets) != 0 {
+			curvirfinish := queue.VirtualFinish(0)
+			if curvirfinish < minvirfinish {
+				minvirfinish = curvirfinish
+				minqueue = queue
+				minidx = q.robinidx
```

Comment:
While `q.robinidx` here is not an incorrect value, proving that requires inspecting the behavior of the `roundrobinqueue()` method; equivalent but easier to understand would be to refer to `idx` here.

Reply:
updated